### DISCUSSION TITLE
image extraction

### DISCUSSION THREAD

**Discussion by pverkind** [pypdfium2-team/pypdfium2#316] [General]
I'm trying to extract images from PDFs by looping over the page objects; if there's only one image object and no text object of the page, I try to extract the image (to retain the original quality of the image); in all other cases, I render the page and save the rendered page as an image (see the code below). 

In some cases, pypdfium fails to extract the image directly (without rendering it); this has to do with pdfium itself, as described here: https://issues.chromium.org/issues/42270939. In those cases, I'm rendering the page and storing the rendered page as an image. 

However, with some PDFs, this doesn't work either, as **the call to pdfium_c.FPDFImageObj_GetBitmap() forces my script to exit instead of returning an error**, as it is designed to do:

```python
# pypdfium2/_helpers/page_objects.py: 

class PdfImage (PdfObject):
    #...
    def get_bitmap(self, render=False):
        """..."""
        
        if render:
            if self.pdf is None:
                raise RuntimeError("Cannot get rendered bitmap of loose page object.")
            raw_bitmap = pdfium_c.FPDFImageObj_GetRenderedBitmap(self.pdf, self.page, self)
        else:
            try:
                raw_bitmap = pdfium_c.FPDFImageObj_GetBitmap(self)  # THIS MAKES THE SCRIPT EXIT INSTEAD OF RAISING AN ERROR
            except:
                raw_bitmap = None
   
        if raw_bitmap is None:    # THIS POINT IS NOT REACHED IF pdfium_c.FPDFImageObj_GetBitmap(self) FAILS
            raise PdfiumError(f"Failed to get bitmap of image {self}.")
        
        return PdfBitmap.from_raw(raw_bitmap)
```

Does anyone have an idea why the code exits instead of raising an error, and how I could solve this?

This is an example of a PDF where this fails: [0309Hallaj.Diwan.pdf](https://github.com/pypdfium2-team/pypdfium2/files/15457407/0309Hallaj.Diwan.pdf)

This is my (slightly simplified) code: 

```python
def extract_images_from_pdf(pdf_folder, pdf_fn, img_folder):
    """Extract images from PDF pages.

    If a page contains a single embedded image and no text objects, the image will be directly extracted;
    in more complicated situations, the page will be rendered and the rendered page saved.

    This should ensure the best quality:size ratio.
    """
    pdf_fp = os.path.join(pdf_folder, pdf_fn)
    outfolder = os.path.join(img_folder, pdf_fn[:-4])
    if not os.path.exists(outfolder):
        os.makedirs(outfolder)

    pdf = pdfium.PdfDocument(pdf_fp)
    
    n_pages = len(pdf)
    for i in range(n_pages):
        page_no = i + 1
        try:
            page = pdf[i]
        except:
            print("ERROR LOADING PAGE", i)
            create_failed_page(outfolder, pdf_fn, page_no)
            continue

        # collect metadata on the objects on the page:
        page_objects = list(page.get_objects())
        image_objects = list(page.get_objects(filter=[3])) # 3 = image objects
        text_objects = list(page.get_objects(filter=[1]))     # 1 = text objects 

        outfn = f"img_{page_no:04d}"
        outfp = os.path.join(outfolder, outfn)

        # extract the page image (directly or after rendering):
        try:
            if len(image_objects) == 1 and len(text_objects) == 0:
                try:
                    image_object.extract(outfp)
                except:
                    print("IMAGE EXTRACTION OF PAGE", page_no, "FAILED. RENDERING PAGE INSTEAD")
                    extract_rendered(page, outfp, scale=2)
            else:
                extract_rendered(page, outfp, scale=2)
        except:
            print("ERROR LOADING PAGE", i)
            create_failed_page(outfp)
```

**Comment by mara004** [comment]
> the call to pdfium_c.FPDFImageObj_GetBitmap() forces my script to exit instead of returning an error

What precisely do you mean by "forces my script to exit"? You need to paste the actual error, otherwise I can't really judge what this may be about. Is it a python traceback, or even a C crash?

However, I have two notes already:
* I recently discovered the pointer checks against None are wrong, and fixed it in the devel branch, but not yet in main. We need to check the boolean value, because a pointer object is always non-None, but it may be a null pointer. So it should be `if not raw_bitmap` instead of `if raw_bitmap is None`. Currently, in case `GetBitmap()` returned a null pointer, the `PdfiumErorr` would not hit, but presumably you would get some later, less descriptive error.
* Note that you have incorrectly altered the `PdfImage` code snippet - try/except wrapping C functions is pointless because they cannot raise a python exception by design (apart from ctypes parameter error if you pass mismatched arguments). Errors are handled through the return value – or, in case of bad C code, the process would just crash.

**Reply by pverkind** [reply]
Thanks for the quick reply!
The problem is exactly that the code doesn't return any error but simply exits without any message. 
(the try... except... was an experiment, apologies for not removing it)
I'll try the development branch code.

**Reply by mara004** [reply]
> The problem is exactly that the code doesn't return any error but simply exits without any message.

That is ... exceedingly weird. I never experienced any such behavior.
We never call `sys.exit()`.
And in case of a C crash I would have expected at least a message like "Segmentation fault (core dumped)".

What OS/terminal are you using?

**Reply by mara004** [reply]
OK, so I tried your sample PDF in a python shell with following code:
```pycon
>>> from pypdfium2 import *
>>> from pypdfium2.raw import *
>>> pdf = PdfDocument("./0309Hallaj.Diwan.pdf")
>>> pdf
<PdfDocument uuid:67c7252d from '.../0309Hallaj.Diwan.pdf'>
>>> for page in pdf:
...     for image in page.get_objects(filter=[FPDF_PAGEOBJ_IMAGE]):
...             image.get_bitmap()
... 
<PdfBitmap uuid:a6984479>
Trace/breakpoint trap (core dumped)
```
So we indeed get something like a C crash. Or rather an assertion failure, it seems.
Apparently, the message just didn't show up on your device, for whatever reason.

**Reply by mara004** [reply]
The first error is happening on page index 1, image index 0. And there are many more where it fails.

**Reply by mara004** [reply]
Further observations:
`pypdfium2 extract-images ~/Downloads/0309Hallaj.Diwan.pdf -o out/ --use-bitmap` obviously fails after the first page.
However, when adding `--render` it actually works, and it turns out the failing images have in common that they contain transparency.
The images without transparency all work – to test, add `--pages 1,36,90,92,114,116`

**Reply by mara004** [reply]
Confirmed via the following code that all other pages fail:
```python
import subprocess

WHITELIST = [p-1 for p in [1,36,90,92,114,116]]

code = """
from pypdfium2 import *
from pypdfium2.raw import *

pdf = PdfDocument("./0309Hallaj.Diwan.pdf")
page = pdf[{i}]
for image in page.get_objects(filter=[FPDF_PAGEOBJ_IMAGE]):
    image.get_bitmap()
"""

for i in range(199):
    print(f"case {i}", end=" ")
    out = subprocess.run(["python3", "-c", code.format(i=i)])
    as_expected = out.returncode == 0 if i in WHITELIST else out.returncode == -5
    print(f"-> returned {out.returncode} (expected? {as_expected})")
    assert as_expected
```

**Reply by mara004** [reply]
To sum up, this is evidently a pdfium issue, and needs reporting there – do you want to do this yourself, or should I?
As a workaround, you can pass `render=True` to `PdfImage.get_bitmap()` on this particular PDF in the meantime.

@pverkind Do you have any more sample files that fail, ideally from a different creator than this one?

**Reply by mara004** [reply]
I went ahead with the bug report: https://issues.chromium.org/issues/343075986

**Comment by pverkind** [comment]
Thank you very much for following up and reporting the bug. 

FYI: I'm on a Windows 11 machine, Python3, and I ran the script in VSCode and IDLE, and on the Windows command line, all with the same result.

I tried with your patch from the devel_new branch, but it produced the same result. 

Good catch that it's the pages with transparency that fail; I also noticed that if I print the ImageNotExtractableError in the `extract_smart` function, I get this message: "Unhandled color space Indexed - don't know how to treat data."

The only other file the script failed with, is from the same creator: [1143CabdGhaniNabulusi.HadraUnsiyya.pdf](https://github.com/pypdfium2-team/pypdfium2/files/15470312/1143CabdGhaniNabulusi.HadraUnsiyya.pdf)

My script is part of a pipeline that should process dozens of texts at a time. Although the problem seems to be with pdfium, it would be great if the pypdfium call to `PdfImage.get_bitmap` (in `_extract_smart`, which itself is called from `PdfImage.extract`), instead of forcing the entire script to exit, would raise an error, which I could catch with a try...except... statement; I could then repeat the call with `render=True`. Do you know of a way to achieve this?

**Reply by mara004** [reply]
> Good catch that it's the pages with transparency that fail; I also noticed that if I print the ImageNotExtractableError in the extract_smart function, I get this message: "Unhandled color space Indexed - don't know how to treat data."

Thanks for the info, I will mention this on the bug report with pdfium. Also thanks for the other sample file.

> My script is part of a pipeline that should process dozens of texts at a time. Although the problem seems to be with pdfium, it would be great if the pypdfium call to PdfImage.get_bitmap (in _extract_smart, which itself is called from PdfImage.extract), instead of forcing the entire script to exit, would raise an error, which I could catch with a try...except... statement; I could then repeat the call with render=True. Do you know of a way to achieve this?

I see the point, but I'm afraid if the C code breaks, that forcibly terminates the process.
So it's not possible to handle this with a simple idiom like try/except.
However, you might be able to create a dedicated worker process to handle pdfium jobs. Then if pdfium crashes, only the worker process would terminate with an error code, but not your main process. In that case, you could show an error message, restart the worker and continue with the next PDF in the pipeline.
That said, depending on your use case, this might be overkill. If it is only about these 2 pdfs and the problem doesn't reoccur, I would perhaps just manually blacklist them to use `render=True`. Unless you are hosting a service that should be able to run unattendedly or something.

**Reply by pverkind** [reply]
Thanks for the advice! I'll try the worker process route - although I hope the Pdfium developers will be able to solve the problem on their end.

**Reply by mara004** [reply]
FYI, upstream has created a [patch](https://pdfium-review.googlesource.com/c/pdfium/+/119970) that will supposedly fix the issue.
Though it will yet take some time until it gets merged and released in pdfium-binaries/pypdfium2.

**Reply by mara004** [reply]
@pverkind A new pdfium-binaries version with the fix is now available:
https://github.com/bblanchon/pdfium-binaries/releases/tag/chromium/6531
I think you can do e.g.
```bash
git clone https://github.com/pypdfium2-team/pypdfium2.git
cd pypdfium2/
git checkout stable
python3 -m pip install .
```
to install with that version, until we release new wheels.

**Comment by pverkind** [comment]
Brilliant, it worked! My pipeline now works without the annoying exit. 
Thank you so much for fixing this so swiftly and following up!
